### ISSUE TITLE
apply_to_mask_decoder doesn’t work？

### ISSUE THREAD

**Issue by luo13766302152-cloud** [Sompote/sam3_lora#28]
Hello, thank you so much for your open‑source work!
When I tried to implement it, I found that apply_to_mask_decoder doesn’t work – the trainable parameters remain unchanged.I then checked the SAM3 image model and saw that it uses a segmentation head.I’m not sure how to resolve this issue.
Did you encounter this problem during your implementation? If so, how did you fix it?Thanks a lot!
Below are my results:The number of trainable parameters remains 4,177,920 (0.49%)/190 regardless of whether apply_to_mask_decoder is enabled.
When set to true, the terminal shows the exact same parameter count. I was confused by this, so I halted the training.
[
  {
    "mAP": 0.16836072932624024,
    "mAP50": 0.2057071541288017,
    "mAP75": 0.16566180427566563,
    "cgF1": 0.0,
    "cgF1_50": 0.0,
    "cgF1_75": 0.0,
    "num_images": 165,
    "num_predictions": 171,
    "prob_threshold": 0.3,
    "nms_iou": 0.7,
    "merge_cracks": false,
    "merge_iou": null,
    "weights_path": "/root/autodl-tmp/outputs/vb_dm_lora/best_lora_weights.pt",
    "lora_config": {
      "rank": 16,
      "alpha": 32,
      "dropout": 0.1,
      "target_modules": [
        "q_proj",
        "k_proj",
        "v_proj",
        "out_proj",
        "qkv",
        "proj"
      ],
      "apply_to_vision_encoder": true,
      "apply_to_text_encoder": false,
      "apply_to_geometry_encoder": false,
      "apply_to_detr_encoder": true,
      "apply_to_detr_decoder": true,
      "apply_to_mask_decoder": false
    },
    "parameters": {
      "total_parameters": 844687670,
      "trainable_parameters": 4177920,
      "trainable_percentage": 0.4946
    },
    "timestamp": "2026-03-30 09:26:59"
  }
]

**Comment by Sompote** [comment]
Thank you for the careful report — you found a real bug, and your diagnosis was exactly right. It's now fixed on main (commit 7a49cb0).

**Root cause**

The component filter in `apply_lora_to_model` (`lora_layers.py`) gated the mask decoder on the substring `"mask_decoder"`:

```python
if "mask_decoder" in module_name and not config.apply_to_mask_decoder:
    return False
```

But as you observed, the SAM3 image model has **no module named `mask_decoder`** — its mask module is `segmentation_head` (a `UniversalSegmentationHead`). So the check never matched anything and the flag was a silent no-op, which is why your trainable parameter count was identical (4,177,920 / 190 modules) with the flag on or off.

There was also a flip side: because `segmentation_head.*` matched *none* of the component gates, the head's prompt cross-attention (`segmentation_head.cross_attend_prompt`) was **always** LoRA-adapted, regardless of the setting — it was already inside your 4,177,920.

**The fix (already pushed)**

The gate now matches `segmentation_head` as well, so the flag works in both directions. After `git pull`, with your config (rank 16):

- `apply_to_mask_decoder: true` → **+32,768** trainable params vs. false (the head's q/k/v/out cross-attention projections, 4 × 2 × 256 × 16)
- `apply_to_mask_decoder: false` → the head is now genuinely frozen, so you'll see **slightly fewer** trainable params than your previous runs (4,177,920 − 32,768 = 4,145,152)

**Expectation to set**

Even when enabled, the delta is modest by design: LoRA wraps `nn.Linear` layers matching `target_modules`, and the only such layers in the segmentation head are the cross-attention projections. The rest of the head is `Conv2d` layers (pixel decoder, seg heads) and an MLP whose inner layer names don't match — those aren't LoRA targets. Mask-quality adaptation mostly flows through the DETR decoder and vision encoder, so don't expect a large jump from this flag alone.

Please `git pull`, re-check the parameter counts, and let me know if the numbers move as described. (If your cgF1 = 0.0 persists after retraining with the current main — which also includes the prompt-discrimination fixes — feel free to open a separate issue with your eval command and we'll dig into that too.)
